flytepropeller/pkg/controller/nodes/array/handler.go

```diff
@@ -349,56 +407,85 @@ func (a *arrayNodeHandler) Handle(ctx context.Context, nCtx interfaces.NodeExecu
+					gatherOutputsRequest.responseChannel <- struct {
+						literalMap map[string]*idlcore.Literal
+						error
+					}{nil, err}
+					continue
 				}
 
 				if task.CoreTask() != nil && task.CoreTask().Interface != nil && task.CoreTask().Interface.Outputs != nil {
-					outputVariables = task.CoreTask().Interface.Outputs.Variables
+					for name := range task.CoreTask().Interface.Outputs.Variables {
+						outputLiterals[name] = nilLiteral
+					}
 				}
 
-				// append nil literal for all output variables
-				for name := range outputVariables {
-					appendLiteral(name, nilLiteral, outputLiterals, len(arrayNodeState.SubNodePhases.GetItems()))
-				}
+				gatherOutputsRequest.responseChannel <- struct {
+					literalMap map[string]*idlcore.Literal
+					error
+				}{outputLiterals, nil}
 			} else {
 				// initialize subNode reader
-				currentAttempt := uint32(arrayNodeState.SubNodeRetryAttempts.GetItem(i))
-				subDataDir, subOutputDir, err := constructOutputReferences(ctx, nCtx, strconv.Itoa(i), strconv.Itoa(int(currentAttempt)))
+				currentAttempt := int(arrayNodeState.SubNodeRetryAttempts.GetItem(i))
+				subDataDir, subOutputDir, err := constructOutputReferences(ctx, nCtx,
+					strconv.Itoa(i), strconv.Itoa(currentAttempt))
 				if err != nil {
-					return handler.UnknownTransition, err
+					gatherOutputsRequest.responseChannel <- struct {
+						literalMap map[string]*idlcore.Literal
+						error
+					}{nil, err}
+					continue
 				}
 
 				// checkpoint paths are not computed here because this function is only called when writing
 				// existing cached outputs. if this functionality changes this will need to be revisited.
 				outputPaths := ioutils.NewCheckpointRemoteFilePaths(ctx, nCtx.DataStore(), subOutputDir, ioutils.NewRawOutputPaths(ctx, subDataDir), "")
 				reader := ioutils.NewRemoteFileOutputReader(ctx, nCtx.DataStore(), outputPaths, nCtx.MaxDatasetSizeBytes())
 
-				// read outputs
-				outputs, executionErr, err := reader.Read(ctx)
-				if err != nil {
-					return handler.UnknownTransition, err
-				} else if executionErr != nil {
-					return handler.UnknownTransition, errors.Errorf(errors.IllegalStateError, nCtx.NodeID(),
-						"execution error ArrayNode output, bad state: %s", executionErr.String())
-				}
+				gatherOutputsRequest.reader = &reader
+				a.gatherOutputsRequestChannel <- gatherOutputsRequest
+			}
 
-				// copy individual subNode output literals into a collection of output literals
-				for name, literal := range outputs.GetLiterals() {
-					appendLiteral(name, literal, outputLiterals, len(arrayNodeState.SubNodePhases.GetItems()))
-				}
+			gatherOutputsRequests = append(gatherOutputsRequests, gatherOutputsRequest)
+		}
+
+		outputLiterals := make(map[string]*idlcore.Literal)
+		workerErrorCollector := errorcollector.NewErrorMessageCollector()
+		for i, gatherOutputsRequest := range gatherOutputsRequests {
+			outputResponse := <-gatherOutputsRequest.responseChannel
```

Comment:
Reading the code I was wondering why are we creating many "small" channels instead of one channel that all workers send on, ordering can be managed through an index field within the response itself.

The benefit of a single channel is that as soon as a worker frees itself up, it can start executing other nodes... the way it's setup now the worker that picks up the last item will wait for all items ahead of it to finish before it can pickup further work...

Reply:
> the way it's setup now the worker that picks up the last item will wait for all items ahead of it to finish before it can pickup further work

I'm not sure this is accurate. For example:
```
package main

import (
	"fmt"
	"time"
)

type request struct {
	index int
	responseChannel chan struct {
		int
	} 
}

func main() {
	requestCount := 20
	workerCount := 4
	requestChannel := make(chan request)
	
	for i := 0; i < workerCount; i++ {
	go func(){
		for {
			request := <- requestChannel
			
			fmt.Printf("COMPLETED %d\n", request.index)
			request.responseChannel <- struct {
				int
			}{request.index}
		}
	}()
	}
	
	requests := make([]request, 0)
	for i := 0; i < requestCount; i++ {
		request := request{
			index: i,
			responseChannel: make(chan struct{
				int
			}, 1),
		}
		
		requests = append(requests, request)
		requestChannel <- request
	}

	time.Sleep(5 * time.Second)
	
	for i, request := range requests {
		response := <-request.responseChannel
		fmt.Printf("RESPONSE %d %d\n", i, response.int)
	}
}
```

Will run all 20 items (in varying orders) and then read responses. This is because the response channel is length 1, so response writing is non-blocking, rather will be buffered. So when submitting items the request channel has no length so that is blocking, but we will not proceed (with overall execution) until all items are completed so that's not very concerning to me.

My only concern with using a single channel is then we have to size it correctly to ensure non-blocking as you suggested, where rather than using the easy `1` we need to use `max-parallelism` potentially? And frequently allocate a 5k length channel where we use a significant subset. I suppose this is all dependent on the relative overhead of creating 5k 1 length channels vs unused space.

I've seen this model used elsewhere and was excited to use it, but certainly understand a single channel approach as well. If you feel strongly, happy to update. The ultimate hope is to use a similar model for overall node evaluation parallelization and also eventing (with a separate worker pool), so it would be easier to get some consensus for uniformity now.